File: src/ada/comms/rest/routes/admin_projects.py
"""Admin project-management routes: ``GET``/``POST /api/admin/projects``,
``DELETE /api/admin/projects/{id}``, the member CRUD family, and the
per-project CI-bot provision/revoke pair.

User-facing project routes (``GET /api/me``, ``GET /api/projects``) live in
:mod:`~.projects` instead, mounted under the plain ``api`` router.

Extracted from ``create_app``; see ``routes/__init__`` for the pattern.
"""

from __future__ import annotations

import re

from fastapi import APIRouter, Depends, HTTPException, Request, Response
from fastapi.responses import JSONResponse

from ada.config import logger

from .. import auth as auth_module
from .. import db as db_module
from ..auth import User
from .deps import require_pool

router = APIRouter()


def _validate_uuid(value: str, what: str = "id") -> str:
    import uuid as _uuid

    try:
        return str(_uuid.UUID(value))
    except (ValueError, AttributeError, TypeError) as exc:
        raise HTTPException(status_code=400, detail=f"invalid {what}") from exc


@router.get("/projects")
async def admin_projects_list(request: Request) -> JSONResponse:
    pool = require_pool(request)
    return JSONResponse({"projects": await db_module.list_all_projects(pool)})


@router.post("/projects")
async def admin_projects_create(
    request: Request,
    user: User = Depends(auth_module.current_user),
) -> JSONResponse:
    pool = require_pool(request)
    body = await request.json()
    slug = (body.get("slug") or "").strip()
    name = (body.get("name") or "").strip()
    if not slug or not name:
        raise HTTPException(status_code=400, detail="slug and name required")
    # Slug shape: lowercase, alnum + hyphens. Keeps URLs / on-disk
    # prefixes predictable; doesn't otherwise constrain the name.
    if not re.fullmatch(r"[a-z0-9][a-z0-9-]{0,62}", slug):
        raise HTTPException(
            status_code=400,
            detail="slug must be lowercase alnum/hyphens (max 63)",
        )
    try:
        project = await db_module.create_project(pool, slug, name)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    # Auto-add the creator as owner so the new project shows up in
    # their /api/me.scopes immediately. Without this the project is
    # orphaned until an admin manually adds someone — easy to forget,
    # and it leaves the creator unable to push artefacts to the
    # project they just made.
    await db_module.add_project_member(pool, project["id"], user.sub, role="owner")
    project["member_count"] = 1
    return JSONResponse(project, status_code=201)


@router.delete("/projects/{project_id}")
async def admin_projects_archive(
    project_id: str,
    request: Request,
) -> Response:
    pool = require_pool(request)
    pid = _validate_uuid(project_id, "project_id")
    ok = await db_module.archive_project(pool, pid)
    if not ok:
        raise HTTPException(status_code=404, detail="project not found")
    return Response(status_code=204)


@router.get("/projects/{project_id}/members")
async def admin_project_members_list(
    project_id: str,
    request: Request,
) -> JSONResponse:
    pool = require_pool(request)
    pid = _validate_uuid(project_id, "project_id")
    if not await db_module.project_exists(pool, pid):
        raise HTTPException(status_code=404, detail="project not found")
    return JSONResponse({"members": await db_module.list_project_members(pool, pid)})


@router.post("/projects/{project_id}/members")
async def admin_project_members_add(
    project_id: str,
    request: Request,
) -> JSONResponse:
    pool = require_pool(request)
    pid = _validate_uuid(project_id, "project_id")
    body = await request.json()
    sub = (body.get("user_sub") or "").strip()
    role = (body.get("role") or "member").strip() or "member"
    if not sub:
        raise HTTPException(status_code=400, detail="user_sub required")
    if not await db_module.project_exists(pool, pid):
        raise HTTPException(status_code=404, detail="project not found")
    added = await db_module.add_project_member(pool, pid, sub, role)
    return JSONResponse(
        {"user_sub": sub, "role": role, "added": added},
        status_code=201 if added else 200,
    )


@router.delete("/projects/{project_id}/members/{user_sub}")
async def admin_project_members_remove(
    project_id: str,
    user_sub: str,
    request: Request,
) -> Response:
    pool = require_pool(request)
    pid = _validate_uuid(project_id, "project_id")
    ok = await db_module.remove_project_member(pool, pid, user_sub)
    if not ok:
        raise HTTPException(status_code=404, detail="not a member")
    return Response(status_code=204)


# A bot name is one path-safe token. The colon is excluded because it is
# the SEPARATOR: a name containing one could spell another bot's subject
# (`ci:<slug>:a:b`) and quietly take over its tokens and its revocation.
_CI_BOT_NAME_RE = re.compile(r"^[a-z0-9][a-z0-9._-]{0,63}$")


def _ci_bot_identity(slug: str, name: str | None) -> tuple[str, str, str]:
    """``(sub, email, display)`` for a project's CI bot.

    Unnamed is ``ci:<slug>`` -- unchanged, so every token already issued
    under that subject keeps working and keeps being rotated by the same
    call as before. A name appends one more segment.
    """
    if not name:
        return f"ci:{slug}", f"ci+{slug}@bot.local", f"CI Bot: {slug}"
    return f"ci:{slug}:{name}", f"ci+{slug}.{name}@bot.local", f"CI Bot: {slug} / {name}"


async def _ci_bot_request(request: Request, project_id: str) -> tuple[object, str, str, str, str]:
    """Shared prologue: validate, resolve the project, build the identity."""
    pool = require_pool(request)
    pid = _validate_uuid(project_id, "project_id")
    try:
        body = await request.json()
    except Exception:
        body = {}
    if not isinstance(body, dict):
        body = {}
    name = (str(body.get("name") or "")).strip().lower() or None
    if name is not None and not _CI_BOT_NAME_RE.match(name):
        raise HTTPException(
            status_code=400,
            detail=(
                "name must be 1-64 characters of a-z, 0-9, dot, dash or underscore, "
                "starting alphanumeric. A colon is not allowed: it separates the "
                "project from the bot, so a name containing one could spell another "
                "bot's identity."
            ),
        )
    row = await pool.fetchrow(
        "SELECT slug FROM projects WHERE id = $1 AND archived_at IS NULL",
        pid,
    )
    if row is None:
        raise HTTPException(status_code=404, detail="project not found")
    sub, email, display = _ci_bot_identity(row["slug"], name)
    return pool, pid, sub, email, display


@router.post("/projects/{project_id}/ci-bot")
async def admin_provision_ci_bot(
    project_id: str,
    request: Request,
) -> JSONResponse:
    """Provision (or rotate the token of) a CI bot user for a project.

    One-shot: creates the bot user row if missing, ensures it's a
    project member, revokes any prior tokens, and mints a fresh
    30-day CLI bearer. The token is returned exactly once, and
    re-calling ROTATES -- prior tokens for that bot stop validating
    immediately via the per-user revoke cutoff. Always admin-gated.

    MORE THAN ONE BOT PER PROJECT. Body: ``{"name": "..."}``, optional.
    Without it the subject is ``ci:<slug>``, exactly as before. With it,
    ``ci:<slug>:<name>``.

    The reason is that one identity per project forces every consumer to
    share one credential, and the revoke cutoff is stored per SUBJECT --
    so rotating for one consumer silently breaks the others, and every
    audit row reads ``ci:<slug>`` no matter which of them acted. Give the
    build uploader and a data-recording worker a name each and both
    problems go away: separate rotation, separate revocation, and an audit
    trail that says which one did the thing.

    Nothing about the token or the revocation model changes to allow it. A
    distinct subject simply HAS its own cutoff, which is why this is a new
    segment on the subject rather than a token id and a revocation list.
    """
    pool, pid, bot_sub, bot_email, bot_display = await _ci_bot_request(request, project_id)

    await db_module.upsert_user(pool, bot_sub, bot_email, bot_display)
    await db_module.add_project_member(pool, pid, bot_sub, role="ci")

    bot_user = User(
        sub=bot_sub,
        email=bot_email,
        display_name=bot_display,
        groups=frozenset(),
        is_admin=False,
    )
    # Rotate: invalidate any tokens minted before now for this bot,
    # then mint a fresh one. The cutoff is iat-based so the token
    # we're about to mint (with a fresh iat) survives.
    await auth_module.revoke_cli_tokens(pool, bot_user)
    config = request.app.state.auth_config
    token, exp = auth_module.mint_cli_token(bot_user, config)
    return JSONResponse(
        {
            "user_sub": bot_sub,
            "token": token,
            "expires_at": exp,
        },
        status_code=201,
    )


@router.post("/projects/{project_id}/ci-bot/revoke")
async def admin_revoke_ci_bot(
    project_id: str,
    request: Request,
) -> JSONResponse:
    """Kill a CI bot's tokens WITHOUT minting a replacement.

    Until now the only way to invalidate a bot's token was to mint another
    one, which is the wrong move for a leaked credential or a decommissioned
    consumer: it hands you a fresh secret you did not want and leaves the
    bot able to act. Revoking on its own is the thing an operator reaches
    for when something has gone wrong, and it did not exist.

    The bot stays a project member. Removing it is a separate, deliberate
    act (``DELETE /projects/{id}/members/{sub}``) -- and keeping the
    membership means its audit history still resolves to a named principal
    rather than a bare subject nobody can identify later.
    """
    pool, _pid, bot_sub, bot_email, bot_display = await _ci_bot_request(request, project_id)
    bot_user = User(
        sub=bot_sub,
        email=bot_email,
        display_name=bot_display,
        groups=frozenset(),
        is_admin=False,
    )
    revoked_at = await auth_module.revoke_cli_tokens(pool, bot_user)
    logger.info("admin: revoked CI bot tokens for %s", bot_sub)
    return JSONResponse({"user_sub": bot_sub, "revoked_at": revoked_at})

File: src/frontend/src/services/api/adminProjects.ts
// Admin projects: project CRUD, membership, CI-bot provisioning, job
// cancellation, and the CLI token pair issued for scripted access.

import { runtime } from "@/runtime/config";

import { ApiError, authedFetch, jsonOrThrow, readDetail } from "./client";

export interface AdminProject {
  id: string;
  slug: string;
  name: string;
  created_at: string | null;
  archived_at: string | null;
  member_count: number;
}

export interface ProjectMember {
  user_sub: string;
  role: string;
  added_at: string | null;
  email: string | null;
  display_name: string | null;
  last_seen_at: string | null;
}

export const adminProjectsApi = {
  /** Mint a 30-day bearer for CLI / pixi-task use. Returned once;
   * the server does not persist it. */
  async adminMintCliToken(): Promise<{ token: string; expires_at: number }> {
    const r = await authedFetch(`${runtime.apiBase()}/admin/auth/cli-token`, {
      method: "POST",
    });
    return jsonOrThrow(r, "adminMintCliToken");
  },

  /** Revoke every previously-minted CLI token for the current user
   * by bumping the per-user cutoff. The OIDC bearer used for this
   * request stays valid. */
  async adminRevokeCliTokens(): Promise<{ revoked_at: number }> {
    const r = await authedFetch(
      `${runtime.apiBase()}/admin/auth/cli-token/revoke`,
      { method: "POST" },
    );
    return jsonOrThrow(r, "adminRevokeCliTokens");
  },

  async adminListProjects(): Promise<AdminProject[]> {
    const r = await authedFetch(`${runtime.apiBase()}/admin/projects`);
    const body = await jsonOrThrow<{ projects: AdminProject[] }>(
      r,
      "adminListProjects",
    );
    return body.projects;
  },

  async adminCreateProject(slug: string, name: string): Promise<AdminProject> {
    const r = await authedFetch(`${runtime.apiBase()}/admin/projects`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ slug, name }),
    });
    return jsonOrThrow<AdminProject>(r, "adminCreateProject");
  },

  /** Provision (or rotate the token of) a synthetic CI bot user for a
   * project. Returns the bearer exactly once — the server does not
   * persist it. Re-calling rotates: the per-user revoke cutoff is
   * bumped before the new token is minted, so any tokens issued
   * previously to that bot stop validating.
   *
   * `name` gives the project more than one bot — `ci:<slug>:<name>`
   * instead of `ci:<slug>`. Without it the subject is unchanged, so
   * an existing bot keeps its identity and its tokens. Use a name per
   * consumer: the revoke cutoff is stored per subject, so consumers
   * sharing one bot cannot be rotated independently, and every audit
   * row reads the same subject whichever of them acted. */
  async adminProvisionCiBot(
    projectId: string,
    name?: string,
  ): Promise<{ user_sub: string; token: string; expires_at: number }> {
    const r = await authedFetch(
      `${runtime.apiBase()}/admin/projects/${encodeURIComponent(projectId)}/ci-bot`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(name ? { name } : {}),
      },
    );
    return jsonOrThrow(r, "adminProvisionCiBot");
  },

  /** Invalidate a CI bot's tokens without minting a replacement.
   *
   * What you want for a leaked credential or a retired consumer:
   * rotating would hand you a fresh secret you did not ask for and
   * leave the bot able to act. The bot stays a project member — remove
   * it separately, so its audit history still resolves to a named
   * principal. */
  async adminRevokeCiBot(
    projectId: string,
    name?: string,
  ): Promise<{ user_sub: string; revoked_at: number }> {
    const r = await authedFetch(
      `${runtime.apiBase()}/admin/projects/${encodeURIComponent(projectId)}/ci-bot/revoke`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(name ? { name } : {}),
      },
    );
    return jsonOrThrow(r, "adminRevokeCiBot");
  },

  /** Cancel and clear any job, whoever started it. Admin only.
   *
   * For a job nothing will ever finish — queued against a capability no
   * live worker serves, or left behind by a retired pool. The user-facing
   * my-jobs cancel filters on the job's owner, so an operator cleaning up
   * after someone else (or after a pool) cannot use it.
   *
   * `cancelled` and `purged` are independent: the audit row and the KV
   * entry can be stuck separately, and a job can need clearing from either
   * or both. */
  async adminCancelJob(
    jobId: string,
  ): Promise<{ job_id: string; cancelled: boolean; purged: boolean }> {
    const r = await authedFetch(
      `${runtime.apiBase()}/admin/jobs/${encodeURIComponent(jobId)}/cancel`,
      { method: "POST" },
    );
    return jsonOrThrow(r, "adminCancelJob");
  },

  async adminArchiveProject(projectId: string): Promise<void> {
    const r = await authedFetch(
      `${runtime.apiBase()}/admin/projects/${encodeURIComponent(projectId)}`,
      { method: "DELETE" },
    );
    if (!r.ok && r.status !== 204) {
      throw new ApiError(
        `adminArchiveProject failed: ${r.status}`,
        r.status,
        await readDetail(r),
      );
    }
  },

  async adminListMembers(projectId: string): Promise<ProjectMember[]> {
    const r = await authedFetch(
      `${runtime.apiBase()}/admin/projects/${encodeURIComponent(projectId)}/members`,
    );
    const body = await jsonOrThrow<{ members: ProjectMember[] }>(
      r,
      "adminListMembers",
    );
    return body.members;
  },

  async adminAddMember(
    projectId: string,
    userSub: string,
    role: string = "member",
  ): Promise<{ user_sub: string; role: string; added: boolean }> {
    const r = await authedFetch(
      `${runtime.apiBase()}/admin/projects/${encodeURIComponent(projectId)}/members`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ user_sub: userSub, role }),
      },
    );
    return jsonOrThrow(r, "adminAddMember");
  },

  async adminRemoveMember(projectId: string, userSub: string): Promise<void> {
    const r = await authedFetch(
      `${runtime.apiBase()}/admin/projects/${encodeURIComponent(projectId)}` +
        `/members/${encodeURIComponent(userSub)}`,
      { method: "DELETE" },
    );
    if (!r.ok && r.status !== 204) {
      throw new ApiError(
        `adminRemoveMember failed: ${r.status}`,
        r.status,
        await readDetail(r),
      );
    }
  },
};
